api/test/wfl/integration/datarepo_test.clj

```diff
@@ -77,20 +78,86 @@
-                     :rows flatten)]
-    (testing "creating snapshot"
-      (fixtures/with-temporary-snapshot
-        (snapshots/unique-snapshot-request tdr-profile dataset table row-ids)
-        #(let [snapshot (datarepo/snapshot %)]
-           (is (= % (:id snapshot))))))))
+        dataset     (datarepo/dataset testing-dataset)
+        table       "flowcell"
+        from        "2021-03-30"
+        until       "2021-03-31"
+        row-ids     (-> (datarepo/query-table-between
+                         dataset
+                         table
+                         [from until]
+                         [:datarepo_row_id])
+                        :rows
+                        flatten)]
+    (fixtures/with-temporary-snapshot
+      (snapshots/unique-snapshot-request tdr-profile dataset table row-ids)
+      #(let [snapshot (datarepo/snapshot %)]
+         (is (= % (:id snapshot)))))))
+
+(deftest test-flattened-query-result
+  (let [samplesheets (-> (datarepo/dataset testing-dataset)
+                         (datarepo/query-table "flowcell" [:samplesheets])
+                         :rows
+                         (->> (mapcat first)))]
+    (is (every? string? samplesheets) "Nested arrays were not normalized.")))
+
+(defn ^:private table->map-view
+  "Create a hash-map 'view' of the BigQuery `table`. A 'view' is one"
+  [{:keys [schema rows] :as _table}]
+  (let [make-entry (fn [idx field] [(-> field :name keyword) idx])
+        key->index (into {} (map-indexed make-entry (:fields schema)))]
+    (map (util/curry #(when-let [idx (key->index %2)] (%1 idx))) rows)))
+
+(defn ^:private maps->table
+  "Transform a list of maps into a table with columns given by `attributes`."
+  [maps attributes]
+  (let [table {:schema {:fields (mapv #(-> {:name (name %)}) attributes)}}
+        f     #(reduce (fn [row attr] (conj row (% attr))) [] attributes)]
+    (assoc table :rows (map f maps))))
+
+(defn ^:private make-entity-import-request-tsv
+  [from-dataset primary-key maps]
+  ;; the columns need to be ordered such that that primary key comes first
+  (let [columns (>>> #(dissoc % primary-key) keys sort #(cons primary-key %))]
+    (-> (map #(datasets/rename-gather % from-dataset) maps)
+        (maps->table (columns from-dataset))
+        bigquery/dump-table->tsv)))
+
+(defn import-table
+  "Import the BigQuery `table` into the `entity` in the Terra `workspace`,
+   using `from-snapshot` to map column names in `snapshot` to the names in the
+   workspace `table`. Return `[entity name]` pairs of the entities imported
+   into the workspace."
+  [table workspace entity primary-key from-snapshot]
+  (let [maps (table->map-view table)]
+    (->> (make-entity-import-request-tsv from-snapshot primary-key maps)
+         .getBytes
+         (firecloud/import-entities workspace))
+    (map (comp #(-> [entity %]) #(% primary-key)) maps)))
+
+;; not sure where this should live!
+(def ^:private snapshot-id "7cb392d8-949b-419d-b40b-d039617d2fc7")
```

Comment:
I think it could just live here!

Reply:
Oh that was meant to apply to the import table functionality. I think we can probably move it into the new workload when we write it.